### PR TITLE
Added Nomic moe v2 embedding model class definition and hf conversion for it

### PR DESCRIPTION
**Pull request by ncylich**
# Add Nomic Embedding Model Support

## Summary
Adds foundational support for Nomic BERT embedding models (e.g., nomic-embed-text-v2-moe) 
with weight loading and HuggingFace conversion. Model architecture is defined but forward 
pass implementation is deferred to future PRs.

### Model Architecture
- New `NomicModel` class with weight loading for:
  - Embedding layer normalization
  - Transformer layers (Q/K/V attention, FFN, layer norms)
  - Mixture-of-Experts (MoE) layers
- Placeholder methods for attention, MLP, and forward pass (to be implemented)

### Configuration
- Added MoE config parameters: `num_experts`, `num_shared_experts`, `num_top_experts`, 
  `moe_every_n_layers`
- New `ModelType::NOMIC` enum and factory support

### HuggingFace Conversion
- Enhanced `convert_hf.py` for Nomic BERT models:
  - Fuses word + token type embeddings
  - Splits combined QKV weight matrices
  - Exports MoE weights (router, per-expert MLPs)
  - Handles embedding layer norms
- Fallback to `AutoModel` for non-causal models
- Improved quantization handling for embeddings/norms/biases

## Files Changed
- `cactus/models/model.h` (+62 lines)
- `cactus/models/model_nomic.cpp` (+82 lines, new)
- `cactus/engine/engine.h` (+6 lines)
- `cactus/engine/engine_model.cpp` (+7 lines)
- `tools/convert_hf.py` (+110/-22 lines)

## Next Steps
Add operation and kernel updates in `origin/moe-ops-for-nomic-embed` and forward pass implementation is complete on the `implemented-nomic-model` branch

### CODE REVIEW

File: cactus/models/model.h [hunk 1]
```diff
@@ -92,5 +92,67 @@ class GemmaModel : public Model {
...
 
+
+class NomicModel : public Model {
+public:
+    NomicModel();
+    explicit NomicModel(const Config& config);
+    ~NomicModel() override = default;
+
+protected:
+    // use_cache must always be false for Nomic
```
Comment (HenryNdubuaku) [inline cactus/models/model.h]: Remove comments

File: cactus/models/model.h [hunk 2]
```diff
@@ -92,5 +92,67 @@ class GemmaModel : public Model {
...
+
+    size_t build_moe_mlp(CactusGraph* gb, size_t normalized_h, uint32_t layer_idx,
+                        ComputeBackend backend) const;
+
+    struct WeightNodeIDs {
+        size_t embedding_layernorm_weight;
+        size_t embedding_layernorm_bias;
+
+        struct LayerWeights {
+            // standard weights
```
Comment (HenryNdubuaku) [inline cactus/models/model.h]: remove comments

File: tools/convert_hf.py [hunk 3]
```diff
@@ -191,26 +194,40 @@ def convert_hf_model_weights(model, output_dir, precision='INT8', args=None):
...
-    embedding_tensor = None
     for name in embed_names:
         if name in state_dict:
             embedding_tensor = state_dict[name]
             save_tensor_with_header(embedding_tensor, output_dir / "token_embeddings.weights", precision, transpose=False, stats_tracker=quantization_stats, args=args, model_type=detected_model_type)
             embedding_found = True
             break
+    if model_type_str == 'nomic_bert':
+        if 'embeddings.word_embeddings.weight' in state_dict:
+            # Nomic BERT has two separate embedding tensors for word and token type, but since there's only 1 token type (shape is [1, 768]), we can add it to the word embedding tensor
```
Comment (HenryNdubuaku) [inline tools/convert_hf.py]: Clean up code comments

File: tools/convert_hf.py [hunk 4]
```diff
@@ -191,26 +194,40 @@ def convert_hf_model_weights(model, output_dir, precision='INT8', args=None):
...
+            fused_embedding_tensor = state_dict['embeddings.word_embeddings.weight'] + state_dict.get('embeddings.token_type_embeddings.weight', torch.zeros([1]))
+            save_tensor_with_header(fused_embedding_tensor, output_dir / "token_embeddings.weights", precision, transpose=False, stats_tracker=quantization_stats, args=args, model_type=detected_model_type)
+            embedding_found = True
     
-    if not embedding_found:
-        pass
+    if embedding_found:
+        # Nomic BERT also has a layer norm after the embedding
```
Comment (HenryNdubuaku) [inline tools/convert_hf.py]: Again remove comments

File: tools/convert_hf.py [hunk 5]
```diff
@@ -256,6 +273,24 @@ def convert_hf_model_weights(model, output_dir, precision='INT8', args=None):
...
+            (['attn.out_proj.weight'], precision, f'layer_{i}_attn_output.weights', False),
+            (['mlp.fc1.bias'], precision, f'layer_{i}_mlp_fc1.bias', False),
+            (['mlp.fc1.weight'], precision, f'layer_{i}_mlp_fc1.weights', False),
+            (['mlp.fc2.bias'], precision, f'layer_{i}_mlp_fc2.bias', False),
+            (['mlp.fc2.weight'], precision, f'layer_{i}_mlp_fc2.weights', False),
+            (['norm1.bias'], precision, f'layer_{i}_norm1.bias', False),
+            (['norm1.weight'], precision, f'layer_{i}_norm1.weights', False),
+            (['norm2.bias'], precision, f'layer_{i}_norm2.bias', False),
+            (['norm2.weight'], precision, f'layer_{i}_norm2.weights', False),
+            # MoE specific weight matrices for Nomic
```
Comment (HenryNdubuaku) [inline tools/convert_hf.py]: Remove coments

File: tools/convert_hf.py [hunk 6]
```diff
@@ -264,6 +299,31 @@ def convert_hf_model_weights(model, output_dir, precision='INT8', args=None):
...
+                        elif tensor.ndim == 2:
+                            tensor = tensor.reshape(3, -1, tensor.size(-1))
+                        else:
+                            raise ValueError(f"Invalid tensor shape: {tensor.shape}")
+                        for j, ch in enumerate(['q', 'k', 'v']):
+                            channel_output_name = output_name.replace('{channel}', ch)
+                            save_tensor_with_header(tensor[j], output_dir / channel_output_name, tensor_precision, transpose=should_transpose, stats_tracker=quantization_stats, args=args, model_type=detected_model_type)
+                        found = True
+                        break
+                    # Handle MoE expert weights - split and save each expert individually
```
Comment (HenryNdubuaku) [inline tools/convert_hf.py]: Remove comments

File: tools/convert_hf.py [hunk 7]
```diff
@@ -264,6 +299,31 @@ def convert_hf_model_weights(model, output_dir, precision='INT8', args=None):
...
+                    elif model_type_str == 'nomic_bert' and pattern.startswith('mlp.experts.') and 'bias' not in pattern:
+                        # For expert weights, the first dimension is the number of experts
```
Comment (HenryNdubuaku) [inline tools/convert_hf.py]: Remove comments

File: tools/convert_hf.py [hunk 8]
```diff
@@ -551,12 +610,21 @@ def convert_hf_to_cactus(model_name, output_dir, precision='INT8', cache_dir=Non
...
-        )
+        try:
+            model = AutoModelForCausalLM.from_pretrained(
+                model_name,
+                cache_dir=cache_dir,
+                trust_remote_code=True,
+                dtype=torch.float32
+            )
+        except ValueError:
+            # Fallback for non-causal models, like embedding models
```
Comment (HenryNdubuaku) [inline tools/convert_hf.py]: Remove comments
